### PR TITLE
add while, break and continue to asm-to-wasm

### PR DESCRIPTION
**Pull request by aseemgarg89**

### CODE REVIEW

File: src/wasm/asm-wasm-builder.cc [hunk 1]
```diff
@@ -117,9 +121,50 @@ class AsmWasmBuilderImpl : public AstVisitor {
...
+  void VisitBreakStatement(BreakStatement* stmt) {
+    DCHECK(in_function_);
+    int i = static_cast<int>(breakable_blocks_.size()) - 1;
+    int jmp_ht = 0;
+    for (; i >= 0; i--) {
+      auto elem = breakable_blocks_.at(i);
+      if (elem.first == stmt->target()) {
+        current_function_builder_->AppendCode(kExprBr, false);
+        if (elem.second) {
+          current_function_builder_->AppendCode(jmp_ht + 1, false);
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Can you extract the target of the jump (i.e. jmp_ht or jmp_ht + 1) out to a local variable?

File: src/wasm/asm-wasm-builder.cc [hunk 2]
```diff
@@ -117,9 +121,50 @@ class AsmWasmBuilderImpl : public AstVisitor {
     }
   }
 
-  void VisitContinueStatement(ContinueStatement* stmt) {}
+  void VisitContinueStatement(ContinueStatement* stmt) {
+    DCHECK(in_function_);
+    int i = static_cast<int>(breakable_blocks_.size()) - 1;
+    int jmp_ht = 0;
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Can we call this "block_distance" or something similar?

File: src/wasm/asm-wasm-builder.cc [hunk 3]
```diff
@@ -412,6 +475,11 @@ class AsmWasmBuilderImpl : public AstVisitor {
     RECURSE(Visit(expr->right()));
   }
 
+#undef CASE
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Can we also give these macros more unique names?

File: src/wasm/asm-wasm-builder.cc (hunk 1 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: done

File: src/wasm/asm-wasm-builder.cc (hunk 2 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: done

File: test/mjsunit/wasm/asm-wasm.js [hunk 4]
```diff
@@ -82,3 +83,186 @@ function TestReturnInBlock() {
...
+    var x = 0;
+    while(x <= 3)
+      x = (x + 1)|0;
+    return x|0;
+  }
+
+  return {caller: caller};
+}
+
+assertEquals(4, WASM.asmCompileRun(TestWhileWithoutBraces.toString()));
```
Comment (titzer) [inline test/mjsunit/wasm/asm-wasm.js]: Can we have some more unique return values than 1, 4, etc?

File: src/wasm/asm-wasm-builder.cc (hunk 3 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: done

File: test/mjsunit/wasm/asm-wasm.js (hunk 4 above)
Comment (aseemgarg89) [inline test/mjsunit/wasm/asm-wasm.js]: done

**Comment by aseemgarg89** [comment]
Commented out the NE and NE_STRICT as the parser ends up creating a EQ/EQ_STRICT node under a unary not node. Will take care of this in future commit

**Comment by titzer** [comment]
lgtm

**Pull request merged**